Firestore/core/src/firebase/firestore/remote/grpc_streaming_reader.cc

```diff
@@ -0,0 +1,136 @@
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using util::AsyncQueue;
+using util::Status;
+
+GrpcStreamingReader::GrpcStreamingReader(
+    std::unique_ptr<grpc::ClientContext> context,
+    std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call,
+    AsyncQueue* worker_queue,
+    const grpc::ByteBuffer& request)
+    : context_{std::move(context)},
+      call_{std::move(call)},
+      worker_queue_{worker_queue},
+      request_{request} {
+}
+
+GrpcStreamingReader::~GrpcStreamingReader() {
+  HARD_ASSERT(!current_completion_,
+              "GrpcStreamingReader is being destroyed without proper shutdown");
+}
+
+void GrpcStreamingReader::Start(CallbackT&& callback) {
+  callback_ = std::move(callback);
+
+  // Coalesce the sending of initial metadata with the first write.
+  context_->set_initial_metadata_corked(true);
+  call_->StartCall(nullptr);
+
+  WriteRequest();
+}
+
+void GrpcStreamingReader::WriteRequest() {
+  SetCompletion([this](const GrpcCompletion* /*ignored*/) { Read(); });
+  *current_completion_->message() = std::move(request_);
+
+  // It is important to indicate to the server that there will be no follow-up
+  // writes; otherwise, the call will never finish.
+  call_->WriteLast(*current_completion_->message(), grpc::WriteOptions{},
+                   current_completion_);
+}
+
+void GrpcStreamingReader::Read() {
+  SetCompletion([this](const GrpcCompletion* completion) {
+    // Accumulate responses
+    responses_.push_back(*completion->message());
+    Read();
+  });
+
+  call_->Read(current_completion_->message(), current_completion_);
+}
+
+void GrpcStreamingReader::Cancel() {
+  if (!current_completion_) {
+    // Nothing to cancel.
+    return;
+  }
+
+  context_->TryCancel();
+  FastFinishCompletion();
+
+  SetCompletion([this](const GrpcCompletion*) {
+    // Deliberately ignored
+  });
+  call_->Finish(current_completion_->status(), current_completion_);
+  FastFinishCompletion();
+}
+
+void GrpcStreamingReader::FastFinishCompletion() {
+  current_completion_->Cancel();
+  // This function blocks.
+  current_completion_->WaitUntilOffQueue();
+  current_completion_ = nullptr;
+}
+
+void GrpcStreamingReader::OnOperationFailed() {
+  // The next read attempt after the server has sent the last response will also
+  // fail; in other words, `OnOperationFailed` will always be invoked, even when
+  // `Finish` will produce a successful status.
+  SetCompletion([this](const GrpcCompletion* completion) {
+    callback_(ConvertStatus(*completion->status()), responses_);
+    // This `GrpcStreamingReader`'s lifetime might have been ended by the
+    // callback.
+  });
+  call_->Finish(current_completion_->status(), current_completion_);
+}
+
+void GrpcStreamingReader::SetCompletion(const OnSuccess& on_success) {
+  // Can't move into lambda until C++14.
+  GrpcCompletion::Callback decorated =
```

Comment:
Thanks for the comment. I think having to potentially fix two different implementations is a very valid concern.

Off the top of my head, I can think of the following:
* leave `GrpcStreamingReader` in place, but make it a wrapper over `GrpcStream`, sort of how `std::queue` is a facade over a general use container;
* remove `GrpcStreamingReader`. The issue is that `GrpcStream` expects that its observer is valid for the stream's lifetime, and that some other class manages it. Without a dedicated wrapper class, `Datastore` would have to do it one way or another;
* one approach to solve the issue from the previous point would be for `Datastore` to define a helper class that stores `GrpcStream` and its observer;
* another approach is for `Datastore` to make itself the observer. In that case, the observer callbacks would need to additionally get the pointer to the stream that invoked the callback as a parameter, so that `Datastore` can distinguish them. `Datastore` would also have to store relevant completions, perhaps indexing them by stream pointer.
* yet another approach would be to make `GrpcStream` accept three `std::function` callbacks instead of a single observer.

I implemented the very first approach, partially because it's the necessary prerequisite to any other approach above (because they would all need `GrpcStream::WriteLast`). Let me know what you think.